lib/fauna/errors.rb

```diff
@@ -91,8 +73,81 @@ class InternalError < FaunaError; end
   # An exception thrown if FaunaDB responds with an HTTP 503.
   class UnavailableError < FaunaError; end
 
+  # :section: ErrorData
+
+  # Data for one error returned by the server.
+  class ErrorData
+    def self.from_hash(hash)
+      code = hash[:code]
+      description = hash[:description]
+      position = ErrorHelpers.map_position hash[:position]
+      failures = hash[:failures].nil? ? nil : hash[:failures].map(&Failure.method(:from_hash))
+      ErrorData.new code, description, position, failures
+    end
+
+    ##
+    # Error code.
+    #
+    # Reference: {FaunaDB Error codes}[https://faunadb.com/documentation#errors]
+    attr_reader :code
+    # Error description.
+    attr_reader :description
+    # Position of the error in a query. May be nil.
+    attr_reader :position
+    # Lit of +Failure+ objects returned by the server. Nil unless code == 'validation failed'.
+    attr_reader :failures
+
+    def initialize(code, description, position, failures)
+      @code = code
+      @description = description
+      @position = position
+      @failures = failures
+    end
+
+    def inspect
+      "ErrorData(#{code.inspect}, #{description.inspect}, #{position.inspect}, #{failures.inspect})"
+    end
+  end
+
   ##
-  # An exception thrown when an unsupported query is used. This currently only applies to
-  # using +object+ within a +databases+ or +keys+ query.
-  class InvalidQuery < FaunaError; end
+  # Part of a +ValidationFailed+.
+  # See the "Invalid Data" section of the {docs}[https://faunadb.com/documentation#errors].
+  class Failure
+    def self.from_hash(hash)
+      Failure.new hash[:code], hash[:description], ErrorHelpers.map_position(hash[:field])
+    end
+
+    # Failure code.
+    attr_reader :code
+    # Failure description.
+    attr_reader :description
+    # Field of the failure in the instance.
+    attr_reader :field
+
+    def initialize(code, description, field)
+      @code = code
+      @description = description
+      @field = field
+    end
+
+    def inspect
+      "Failure(#{code.inspect}, #{description.inspect}, #{field.inspect})"
+    end
+  end
+
+  module ErrorHelpers #:nodoc:
+    def self.map_position(position)
+      if position.nil?
```

Comment:
I think this could actually be simplified down to `unless position.nil?`, as `nil` gets passed back if nothing's returned.